pkg/login/social/connectors/org_role_mapper.go

```diff
@@ -0,0 +1,189 @@
+package connectors
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/org"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+const MapperMatchAllOrgID = -1
+
+type OrgRoleMapper struct {
+	cfg        *setting.Cfg
+	logger     log.Logger
+	orgService org.Service
+}
+
+func ProvideOrgRoleMapper(cfg *setting.Cfg, orgService org.Service) *OrgRoleMapper {
+	return &OrgRoleMapper{
+		cfg:        cfg,
+		logger:     log.New("orgrole.mapper"),
+		orgService: orgService,
+	}
+}
+
+func (m *OrgRoleMapper) MapOrgRoles(ctx context.Context, externalOrgs []string, orgMappingSettings []string, directlyMappedRole org.RoleType) (map[int64]org.RoleType, error) {
+	orgMapping := m.splitOrgMappingSettings(ctx, orgMappingSettings)
+	if len(orgMapping) == 0 {
+		return m.GetDefaultOrgMapping(directlyMappedRole), nil
+	}
+
+	userOrgRoles := getMappedOrgRoles(externalOrgs, orgMapping)
+
+	m.handleGlobalOrgMapping(userOrgRoles)
+
+	if len(userOrgRoles) == 0 {
+		return m.GetDefaultOrgMapping(directlyMappedRole), nil
+	}
+
+	if directlyMappedRole == "" {
+		m.logger.Debug("No direct role mapping found")
+		return userOrgRoles, nil
+	}
+
+	m.logger.Debug("Direct role mapping found", "role", directlyMappedRole)
+
+	// Merge roles from org mapping `org_mapping` with role from direct mapping
+	for orgID, role := range userOrgRoles {
+		userOrgRoles[orgID] = getTopRole(directlyMappedRole, role)
+	}
+
+	return userOrgRoles, nil
+}
+
+func (m *OrgRoleMapper) GetDefaultOrgMapping(directlyMappedRole org.RoleType) map[int64]org.RoleType {
+	orgRoles := make(map[int64]org.RoleType, 0)
+
+	orgID := int64(1)
+	if m.cfg.AutoAssignOrg && m.cfg.AutoAssignOrgId > 0 {
+		orgID = int64(m.cfg.AutoAssignOrgId)
+	}
+
+	if directlyMappedRole == "" || !directlyMappedRole.IsValid() {
+		orgRoles[orgID] = org.RoleType(m.cfg.AutoAssignOrgRole)
+	} else {
+		orgRoles[orgID] = directlyMappedRole
+	}
+
+	return orgRoles
+}
+
+func (m *OrgRoleMapper) handleGlobalOrgMapping(orgRoles map[int64]org.RoleType) {
+	// No global role mapping => return
+	globalRole, ok := orgRoles[MapperMatchAllOrgID]
+	if !ok {
+		return
+	}
+
+	allOrgIDs, err := m.getAllOrgs()
+	if err != nil {
+		// Prevent resetting assignments
+		clear(orgRoles)
+		m.logger.Warn("error fetching all orgs, removing org mapping to prevent org sync")
+		return
+	}
+
+	// Remove the global role mapping
+	delete(orgRoles, MapperMatchAllOrgID)
+
+	// Global mapping => for all orgs get top role mapping
+	for orgID := range allOrgIDs {
+		orgRoles[orgID] = getTopRole(orgRoles[orgID], globalRole)
+	}
+}
+
+func (m *OrgRoleMapper) splitOrgMappingSettings(ctx context.Context, mappings []string) map[string]map[int64]org.RoleType {
```

Comment:
I think this is unnecessary to recompute on every login. We just have to solve these if settings change so it would be better for the providers to perform this logic and store the result and then pass it to the org role mapper

Reply:
Thanks, I'll keep this in mind!

Reply:
Done!